kani-compiler/src/kani_middle/transform/contracts.rs

```diff
@@ -232,8 +228,297 @@ impl AnyModifiesPass {
+    unused_closures: HashSet<ClosureDef>,
+}
+
+impl TransformPass for FunctionWithContractPass {
+    fn transformation_type() -> TransformationType
+    where
+        Self: Sized,
+    {
+        TransformationType::Stubbing
+    }
+
+    fn is_enabled(&self, _query_db: &QueryDb) -> bool
+    where
+        Self: Sized,
+    {
+        true
+    }
+
+    /// Transform the function body by replacing it with the stub body.
+    fn transform(&mut self, tcx: TyCtxt, body: Body, instance: Instance) -> (bool, Body) {
+        trace!(function=?instance.name(), "FunctionWithContractPass::transform");
+        match instance.ty().kind().rigid().unwrap() {
+            RigidTy::FnDef(def, args) => {
+                if let Some(mode) = self.contract_mode(tcx, *def) {
+                    self.mark_unused(tcx, *def, &body, mode);
+                    let new_body = self.set_mode(tcx, body, mode);
+                    (true, new_body)
+                } else if KaniAttributes::for_instance(tcx, instance).fn_marker()
+                    == Some(Symbol::intern("kani_register_contract"))
+                {
+                    let run = Instance::resolve(find_fn_def(tcx, "KaniRunContract").unwrap(), args)
+                        .unwrap();
+                    (true, run.body().unwrap())
+                } else {
+                    // Not a contract annotated function
+                    (false, body)
+                }
+            }
+            RigidTy::Closure(def, _args) => {
+                if self.unused_closures.contains(def) {
+                    // Delete body and mark it as unreachable.
+                    let mut new_body = MutableBody::from(body);
+                    new_body.clear_body(TerminatorKind::Unreachable);
+                    (true, new_body.into())
+                } else {
+                    // Not a contract annotated function
+                    (false, body)
+                }
+            }
+            _ => {
+                /* static variables case */
+                (false, body)
+            }
         }
     }
 }
+
+impl FunctionWithContractPass {
+    /// Build the pass by collecting which functions we are stubbing and which ones we are
+    /// verifying.
+    pub fn new(tcx: TyCtxt, unit: &CodegenUnit) -> FunctionWithContractPass {
+        if let Some(harness) = unit.harnesses.first() {
+            let attrs = KaniAttributes::for_instance(tcx, *harness);
+            let check_fn = attrs.interpret_for_contract_attribute().map(|(_, def_id, _)| def_id);
+            let replace_fns: HashSet<_> = attrs
+                .interpret_stub_verified_attribute()
+                .iter()
+                .map(|(_, def_id, _)| *def_id)
+                .collect();
+            FunctionWithContractPass { check_fn, replace_fns, unused_closures: Default::default() }
+        } else {
+            // Building the model for tests or public functions.
+            FunctionWithContractPass::default()
```

Comment:
I don't understand what you mean here.

Reply:
Ah, I can update this comment.  This is for cases where the reachability mode is not harnesses.